File: Online Judges/Huxley/556Ciclos.py
import sys
sys.setrecursionlimit(2**20)

def hasCycle(graph, visited, start, parent):
    visited[start] = 1
    for u in graph[start]:
        if (visited[u] == 1 and u != parent):
            return(1)
        elif (visited[u] == 0 and hasCycle(graph, visited, u, start)):
            return(1)
    visited[start] = 2
    return(0)

case = 1
while (True):
    try:
        if (case > 1):
            trash = input()
        vertices, edges = map(int, input().split())
    except:
        break
    
    graph = [[] for i in range(vertices)]
    for i in range(edges):
        u, v = map(int, input().split())
        graph[u] += [v]
        graph[v] += [u]

    visited = [0] * vertices
    print(case, hasCycle(graph, visited, u, -1))
    case += 1

File: Online Judges/Huxley/556Ciclos.cpp
#include <bits/stdc++.h>
using namespace std;

int hasCycle(vector<int> graph[], int visited[], int u, int parent)
{
  visited[u] = 1;
  for (auto v: graph[u])
  {
    if (visited[v] == 1 && v != parent) return(1);
    if (!visited[v] && hasCycle(graph, visited, v, u)) return(1);
  }
  visited[u] = 2;
  return(0);
}

int main()
{
  int vertices, edges, now = 1;
  while (scanf("%d %d", &vertices, &edges) != EOF)
  {
    vector<int> graph[vertices];
    int u, v;
    for (int i = 0; i < edges; i ++)
    {
      scanf("%d %d", &u, &v);
      graph[u].push_back(v);
      graph[v].push_back(u);
    }

    int visited[vertices]; memset(visited, 0, sizeof(visited));
    printf("%d %d\n", now ++, hasCycle(graph, visited, 0, -1));
  }
  
  return(0);
}